docs/whitepaper/whitepaper.md

```diff
@@ -0,0 +1,591 @@
+
+Because the features of different ledgers can be very diverse, the plugin interface has feature checks built into allowing callers/client applications to **determine programmatically, at runtime** if a certain feature is supported or not on a given ledger.
+
+```typescript
+export interface LedgerConnector {
+  // method to verify a signature coming from a given ledger that this connector is responsible for connecting to.
+  verifySignature(message, signature): Promise<boolean>;
+
+  // used to call methods on smart contracts or to move assets between wallets
+  transact(transactions: Transaction[]);
+
+  getPermissionScheme(): Promise<PermissionScheme>;
+
+  getTransactionFinality(): Promise<TransactionFinality>;
+
+  addForeignValidator(): Promise<void>;
+}
+
+export enum TransactionFinality {
+  GUARANTEED = "GUARANTEED",
+  NOT_GUARANTEED = "NOT_GUARANTEED
+}
+
+export enum PermissionScheme {
+  PERMISSIONED = "PERMISSIONED",
+  PERMISSIONLESS = "PERMISSIONLESS"
+}
+
+```
+
+### 4.3.2 Identity Federation Plugins
+
+Identity federation plugins operate inside the API Server and need to implement the interface of a common PassportJS Strategy:
+https://github.com/jaredhanson/passport-strategy#implement-authentication
+
+```typescript
+abstract class IdentityFederationPlugin {
+  constructor(options: any): IdentityFederationPlugin;
+  abstract authenticate(req: ExpressRequest, options: any);
+  abstract success(user, info);
+  abstract fail(challenge, status);
+  abstract redirect(url, status);
+  abstract pass();
+  abstract error(err);
+}
+```
+
+#### 4.3.1.1 X.509 Certificate Plugin
+
+The X.509 Certificate plugin facilitates clients authentication by allowing them to present a certificate instead of operating with authentication tokens.
+This technically allows calling clients to assume the identities of the validator nodes through the REST API without having to have access to the signing private key of said validator node.
+
+PassportJS already has plugins written for client certificate validation, but we go one step further with this plugin by providing the option to obtain CA certificates from the validator nodes themselves at runtime.
+
+### 4.3.3 Storage Plugins
+
+Storage plugins allow the higher level packages to store and retrieve configuration metadata for a `BIF` cluster such as:
+* Who are the active validators and what are the hosts where said validators are accessible over a network
+* What public keys belong to which validator nodes
+* What transactions have been scheduled, started, completed
+
+```typescript
+interface StoragePlugin {
+  async get<T>(key: string): Promise<T>;
+  async set<T>(key: string, value: T): Promise<void>;
+}
+```
+
+# 5. Identities, Authentication, Authorization
```

Comment:
This is probably the most difficult thing to get right.  It also might be the easiest thing for us to get started on with respect to merging the codebases.

Reply:
Agreed, as soon as we have sufficient level of concensus on the design I plan to crank out a lot of code for the API server starting with a rudimentary but functional web authentication layer that we can build on top of the rest of the components/plugins dealing with the actual ledgers.